packages/analytics-browser/src/session-manager.ts

```diff
@@ -1,31 +1,67 @@
-import { BrowserConfig } from '@amplitude/analytics-types';
-import { AMPLITUDE_PREFIX } from '@amplitude/analytics-core';
-
-export const updateCookies = (config: BrowserConfig, lastEventTime?: number) => {
-  const cookieName = getCookieName(config.apiKey);
-  config.cookieStorage.set(cookieName, {
-    userId: config.userId,
-    deviceId: config.deviceId,
-    sessionId: config.sessionId,
-    lastEventTime: lastEventTime ?? config.cookieStorage.get(cookieName)?.lastEventTime,
-    optOut: Boolean(config.optOut),
-  });
-};
-
-export const checkSessionExpiry = (config: BrowserConfig) => {
-  const cookieName = getCookieName(config.apiKey);
-  const lastEventTime = config.cookieStorage.get(cookieName)?.lastEventTime;
-  const now = Date.now();
-  if (lastEventTime && now - lastEventTime >= config.sessionTimeout) {
-    config.sessionId = now;
-    updateCookies(config);
-  }
-};
-
-export const getCookieName = (apiKey: string, limit = 10) => {
-  return `${AMPLITUDE_PREFIX}_${apiKey.substring(0, limit)}`;
-};
-
-export const getOldCookieName = (apiKey: string) => {
-  return `${AMPLITUDE_PREFIX.toLowerCase()}_${apiKey.substring(0, 6)}`;
-};
+import {
+  UserSession,
+  Storage,
+  SessionManager as ISessionManager,
+  SessionManagerOptions,
+} from '@amplitude/analytics-types';
+import { getCookieName as getStorageKey } from './utils/cookie-name';
+
+export class SessionManager implements ISessionManager {
+  storageKey: string;
+  sessionTimeout: number;
+  cache: UserSession;
+
+  constructor(private storage: Storage<UserSession>, options: SessionManagerOptions) {
+    this.storageKey = getStorageKey(options.apiKey);
+    this.sessionTimeout = options.sessionTimeout;
+    this.cache = this.storage.get(this.storageKey) ?? {
+      optOut: false,
+    };
+  }
+
+  setSession(session: Partial<UserSession>) {
+    this.cache = { ...this.cache, ...session };
+    this.storage.set(this.storageKey, this.cache);
+  }
+
+  getSessionId() {
+    // Note: always read directly from storage
+    return this.storage.get(this.storageKey)?.sessionId;
+  }
```

Comment:
this is how session id gets reconciled, sdk will always read from cookie storage for session id. so when cookies are updated by a different tab, other tabs should be able to read new session id as soon as it does.

other user session data are cached to cookies are not accessed and parsed more often then needed